File: srcs/srcs/Matrix.cpp
#include "../includes/Matrix.hpp"

//----------------------Constructor/Destructor----------------------//
Matrix::Matrix(size_t rows, size_t columns) : rows(rows), columns(columns) {
    matrix = new double*[rows];

    for (size_t i = 0; i < rows; i++) {
        matrix[i] = new double[columns];
        for (size_t j = 0; j < columns; j++)
            matrix[i][j] = 0;
    }
}

Matrix::~Matrix() {
    for (size_t i = 0; i < this->rows; i++) {
        delete[] this->matrix[i];
    }

    delete[] this->matrix; 
}

//----------------------Maths functions----------------------//
void    Matrix::add(Matrix& mat) {
    if (this->getSize() != mat.getSize())
        throw std::length_error("Error: Matrix have to get same sizes");

    Matrix newMat(this->rows, this->columns);

    for (size_t i = 0; i < this->rows; i++) {
        for (size_t j = 0; j < this->columns; j++) {
            newMat.insert(i, j, (this->operator()(i, j) + mat(i, j)));
        }
    }

    newMat.print();
}

void    Matrix::scl(double scl) {
    Matrix newMat(this->rows, this->columns);

    for (size_t i = 0; i < this->rows; i++) {
        for (size_t j = 0; j < this->columns; j++) {
            newMat.insert(i, j, (this->operator()(i, j) * scl));
        }
    }

    newMat.print();
}

void    Matrix::sub(Matrix& mat) {
    if (this->getSize() != mat.getSize())
        throw std::length_error("Error: Matrix have to get same sizes");

    Matrix newMat(this->rows, this->columns);

    for (size_t i = 0; i < this->rows; i++) {
        for (size_t j = 0; j < this->columns; j++) {
            newMat.insert(i, j, (this->operator()(i, j) - mat(i, j)));
        }
    }

    newMat.print();
}

//----------------------Utils functions----------------------//
void    Matrix::insert(size_t row, size_t column, double val) {
    if (this->rows < row || this->columns < column)
        throw std::length_error("Error: Values have to be in range of the Matrix");
    this->matrix[row][column] = val;
}

bool    Matrix::isSquare() const {
    if (this->rows == this->columns)
        return true;
    return false;
}    

void    Matrix::print() const {
    for (size_t i = 0; i < this->rows; i++) {
        std::cout << "(";
        for (size_t j = 0; j < this->columns; j++) {
            if (j == this->columns - 1)
                std::cout << this->matrix[i][j];
            else
                std::cout << this->matrix[i][j] << ", ";
        }
        std::cout << ")" << std::endl;
    }
}

//----------------------Getters----------------------//
size_t Matrix::getColumns() const {
    return this->columns;
}

size_t Matrix::getRows() const {
    return this->rows;
}

std::pair<size_t, size_t> Matrix::getSize() {
    std::pair<size_t, size_t> pair(this->rows, this->columns);

    return pair;
}
//----------------------Setters----------------------//


//----------------------Overload operators----------------------//
double  Matrix::operator()(size_t row, size_t column) const {
    if (this->rows < row || this->columns < column)
        throw std::length_error("Error: Values have to be in range of the Matrix");
    return this->matrix[row][column];
}

File: Matrix.py
from Vector import Vector

class Matrix:
    '''Matrix class'''

    def __init__(self, mat):
        '''Matrix Constructor'''

        if (not all(isinstance(item, Vector) for item in mat)):
            return print("Error: argument need to be Vector")
        
        if (not all(item.size() == mat[0].size() for item in mat)):
            return print("Error: Vector must have the same size")

        self.mat = mat

    def add(self, oth):
        '''Matrix ADD method'''
    
        if (not isinstance(oth, Matrix)):
            return print("Error: arguments need to be Matrix")
        if (self.size() != oth.size()):
            return print("Error: Matrix must have the same size")
        
        for i in range(self.size()):
            self.mat[i].add(oth.mat[i])

        return self.mat

    def scl(self, scl: float | int):
        '''Matrix SCALAR method'''

        if (not isinstance(scl, (float, int))):
            return print("Error: scalar must be float or integer")
        
        for i in range(self.size()):
            self.mat[i].scl(scl)
        
        return self.mat

    def sub(self, oth):
        '''Matrix SUB method'''

        if (not isinstance(oth, Matrix)):
            return print("Error: arguments need to be Matrix")
        if (self.size() != oth.size()):
            return print("Error: Matrix must have the same size")
        
        for i in range(self.size()):
            self.mat[i].sub(oth.mat[i])

        return self.mat

    def print(self):
        '''Matrix print method'''

        for i in range(self.size()):
            self.mat[i].print()

    def size(self):
        '''Matrix size method'''

        return len(self.mat)
